### PR TITLE
[BugFix] consistent use of non_blocking in tensordict and torch.Tensor on non-cuda devices

### PR DESCRIPTION
**Pull request by ji-huazhong**
## Description

Describe your changes in detail.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/tensordict/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/tensordict/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: tensordict/base.py [hunk 1]
```diff
@@ -14258,14 +14259,17 @@ def _has_mps(self):
...
+
+        if device_type == "cuda":
             # TODO: dynamo doesn't like torch.cuda.is_initialized
             if not is_compiling() and torch.cuda.is_initialized():
                 torch.cuda.synchronize()
-        elif self._has_mps:
-            mps = getattr(torch, "mps", None)
-            if mps is not None:
-                mps.synchronize()
+        else:
```
Comment (vmoens) [inline tensordict/base.py:14270]: Double indent

**Review by vmoens — COMMENTED** [review]
Thanks for this PR!
Can you further explain why this is needed and what problem is solved?
Thanks!

**Comment by ji-huazhong** [comment]
> Thanks for this PR! Can you further explain why this is needed and what problem is solved? Thanks!

Hi @vmoens We have recently been using VERL for post-training our model on the Ascend NPUs. We observed that as the training progresses, the gradient norm (grad norm) consistently diverges (runs out of control). In contrast, all metrics converge well when training on NV GPUs.

Coincidentally, we found that adding synchronization (torch.npu.synchronize) at the connection points between different components in the RL workflow resolved this issue, and it no longer recurs.

In VERL, a single controller distributes data(organized in the form of TensorDict) to various workers (each associated with a device) via Ray for computation. After the computation is completed, `td.to("cpu")` is called to move the data from the device to the CPU (see [here](https://github.com/volcengine/verl/blob/9f0f8b0e7c227bedacd16eacaf662f00d2de4367/verl/workers/fsdp_workers.py#L738)). The data is then aggregated back to the single controller through Ray, which then proceeds to execute the next phase of computation jobs. It was precisely by adding synchronization after `td.to("cpu")` that the gradient norm divergence issue ceased to occur.

Through further investigation, we discovered that the `to` operation of TensorDict appears to be non-blocking. Meanwhile, it ensures that when the target device is specified as `to('cpu')`, the `_sync_all` method is invoked to guarantee that the data transfer from the device to the CPU is successfully completed before the data is used. see: https://github.com/pytorch/tensordict/blob/5e781517a1d638b55c3c250d990921fea9ba3c69/tensordict/base.py#L14130-L14136 

https://github.com/pytorch/tensordict/blob/5e781517a1d638b55c3c250d990921fea9ba3c69/tensordict/base.py#L14260-L14268

However, this guarantee only covers CUDA and MPS devices. PyTorch also supports third-party hardware such as Intel XPU and Ascend NPU, and this PR  aims to extend this guarantee to cover these third-party hardware platforms as well.

**Review by vmoens — APPROVED** [review]
LGTM I like it!

**Comment by FightingZhen** [comment]
Great to see this PR merged! When will the next version containing this feature be released? @vmoens

**Comment by vmoens** [comment]
Hoping to release this by end of week but I have a bunch of CI issues...
